mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/transformation/mpsToModelix/initial/ModelSynchronizer.kt

```diff
@@ -123,83 +155,109 @@ class ModelSynchronizer(private val branch: IBranch, postponeReferenceResolution
 
     private fun addModelImportToCloud(source: SModel, targetModel: SModel) {
         val modelixId = nodeMap[source]!!
-
-        val modelImportsLink = BuiltinLanguages.MPSRepositoryConcepts.Model.modelImports
-        val modelReferenceConcept = BuiltinLanguages.MPSRepositoryConcepts.ModelReference
-
         val cloudParentNode = branch.getNode(modelixId)
-        val cloudModelReference = cloudParentNode.addNewChild(modelImportsLink, -1, modelReferenceConcept)
-
-        nodeMap.put(source, targetModel.reference, cloudModelReference.nodeIdAsLong())
+        val childLink = BuiltinLanguages.MPSRepositoryConcepts.Model.modelImports
 
-        // warning: might be fragile, because we synchronize the fields by hand
+        val targetModelReference = BuiltinLanguages.MPSRepositoryConcepts.ModelReference.model
         val targetModelModelixId = nodeMap[targetModel]!!
         val cloudTargetModel = branch.getNode(targetModelModelixId)
-        cloudModelReference.setReferenceTarget(
-            BuiltinLanguages.MPSRepositoryConcepts.ModelReference.model,
-            cloudTargetModel,
-        )
+        val idProperty = BuiltinLanguages.MPSRepositoryConcepts.Model.id
+        val cloudTargetModelId = cloudTargetModel.getPropertyValue(idProperty)
+
+        // duplicate check and sync
+        val modelImportExists = cloudParentNode.getChildren(childLink).firstOrNull {
+            cloudTargetModelId == it.getReferenceTarget(targetModelReference)?.getPropertyValue(idProperty)
+        } != null
+        if (modelImportExists) {
+            logger.warn { "Model Import for Model ${targetModel.name} from Model ${source.name} will not be synchronized, because it already exists on the server." }
```

Comment:
tip: use early returns to avoid to much nesting.

Reply:
fixed in 54ed97610d6dc99daceab34dfa51119ad9c8cb56